### PR TITLE
tensorflow probability examples

### PR DESCRIPTION
**Pull request by dizcology**
1. add tools for adding and modifying examples from tensorflow/probability to be runnable on CMLE

1. add bnn example

1. add grammar_var example

more examples will be added in a separate PR after this PR has been reviewed and approved.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/cloudml-samples/384)

### CODE REVIEW

File: tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
+# Copyright 2018 The TensorFlow Probability Authors.
```
Comment (happyhuman) [inline tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py]: 2019?

File: tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+matplotlib.use("Agg")
+from matplotlib import figure  # pylint: disable=g-import-not-at-top
+from matplotlib.backends import backend_agg
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+from tensorflow.contrib.learn.python.learn.datasets import mnist
+
+# TODO(b/78137893): Integration tests currently fail with seaborn imports.
```
Comment (happyhuman) [inline tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py]: Should we remove this TODO?

File: tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+  # and validation inputs.
+  handle = tf.placeholder(tf.string, shape=[])
+  feedable_iterator = tf.data.Iterator.from_string_handle(
+      handle, training_batches.output_types, training_batches.output_shapes)
+  images, labels = feedable_iterator.get_next()
+
+  return images, labels, handle, training_iterator, heldout_iterator
+
+
+def build_fake_data(num_examples=10):
```
Comment (happyhuman) [inline tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py]: Should we move this function to another python file that contains testing stuff?

File: tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+  mnist_data.validation.images = np.float32(np.random.randn(
+      num_examples, *IMAGE_SHAPE))
+  mnist_data.validation.labels = np.int32(np.random.permutation(
+      np.arange(num_examples)))
+  mnist_data.validation.num_examples = num_examples
+  return mnist_data
+
+
+def main(argv):
+  del argv  # unused
```
Comment (happyhuman) [inline tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py]: Remove this line? If argv is unused, why pass it to the function?

File: tensorflow_probability/grammar_vae/trainer/grammar_vae.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
+# Copyright 2018 The TensorFlow Probability Authors.
```
Comment (happyhuman) [inline tensorflow_probability/grammar_vae/trainer/grammar_vae.py]: 2019?

File: tensorflow_probability/grammar_vae/trainer/grammar_vae.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+    """Runs the model forward to generate a sequence of productions.
+
+    Args:
+      inputs: Unused.
+
+    Returns:
+      productions: Tensor of shape [1, num_productions, num_production_rules].
+        Slices along the `num_productions` dimension represent one-hot vectors.
+    """
+    del inputs  # unused
```
Comment (happyhuman) [inline tensorflow_probability/grammar_vae/trainer/grammar_vae.py]: Why pass inputs if it is to be deleted?

File: tensorflow_probability/grammar_vae/trainer/grammar_vae.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+    """Sets random variable values to its aligned value."""
+    name = kwargs.get("name")
+    if name in model_kwargs:
+      kwargs["value"] = model_kwargs[name]
+    return ed.interceptable(f)(*args, **kwargs)
+  return set_values
+
+
+def main(argv):
+  del argv  # unused
```
Comment (happyhuman) [inline tensorflow_probability/grammar_vae/trainer/grammar_vae.py]: Same as above

File: tensorflow_probability/grammar_vae/trainer/tfgfile_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    # When an example writes to local disk, change it to write to GCS.
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+        fname = kwargs.pop('fname')
+        with tf.gfile.GFile(fname, 'w') as fobj:
+            kwargs['fname'] = fobj
+            return_value = f(*args, **kwargs)
+        return return_value
+
+    return wrapper
```
Comment (happyhuman) [inline tensorflow_probability/grammar_vae/trainer/tfgfile_wrapper.py]: Add a new line at the end of this file?

File: tensorflow_probability/tools/cmle_package.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+import os
+import shutil
+import urllib2
+
+SOURCE_BASE = "https://raw.githubusercontent.com/tensorflow/probability/{branch}/tensorflow_probability/examples/{source_name}"
+BRANCH = 'r0.5'
+
+
+class CMLEPackage(object):
+    def __init__(self, source_name, requires=None, tfgfile_wrap=None):
```
Comment (happyhuman) [inline tensorflow_probability/tools/cmle_package.py]: This file uses 4 spaces for indentation. There are other python files above which use 2 spaces for indentation. Let's make it consistent (I'd go with 4 spaces).

File: tensorflow_probability/grammar_vae/trainer/tfgfile_wrapper.py (hunk 8 above)
Comment (dizcology) [reply tensorflow_probability/grammar_vae/trainer/tfgfile_wrapper.py]: done

File: tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py (hunk 2 above)
Comment (gogasca) [reply tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py]: TF is moving out of contrib, what about: tensorflow.keras.datasets

File: tensorflow_probability/README.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
+# TensorFlow Probability Examples on Cloud Machine Learning Engine.
```
Comment (gogasca) [inline tensorflow_probability/README.md]: Rename to AI platform

File: tensorflow_probability/bayesian_neural_network/submit.sh [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+JOB_DIR=$BUCKET"tfp/"$JOB_NAME"/"
+
+gcloud ml-engine jobs submit training $JOB_NAME \
+    --job-dir $JOB_DIR  \
+    --package-path $TRAINER_PACKAGE_PATH \
+    --module-name $MAIN_TRAINER_MODULE \
+    --region us-central1 \
+    --config config.yaml \
+    --runtime-version 1.13 \
+    --python-version 2.7 \
```
Comment (gogasca) [inline tensorflow_probability/bayesian_neural_network/submit.sh]: nit: Does it work with Python 3.x ?

File: tensorflow_probability/grammar_vae/submit.sh [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+JOB_DIR=$BUCKET"tfp/"$JOB_NAME"/"
+
+gcloud ml-engine jobs submit training $JOB_NAME \
+    --job-dir $JOB_DIR  \
+    --package-path $TRAINER_PACKAGE_PATH \
+    --module-name $MAIN_TRAINER_MODULE \
+    --region us-central1 \
+    --config config.yaml \
+    --runtime-version 1.13 \
+    --python-version 2.7 \
```
Comment (gogasca) [inline tensorflow_probability/grammar_vae/submit.sh]: nit: same as above:

File: tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+
+#### References
+
+[1]: Yann LeCun, Leon Bottou, Yoshua Bengio, and Patrick Haffner.
+     Gradient-based learning applied to document recognition.
+     _Proceedings of the IEEE_, 1998.
+     http://yann.lecun.com/exdb/publis/pdf/lecun-01a.pdf
+"""
+
+from __future__ import absolute_import
```
Comment (gogasca) [inline tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py]: Import order:

```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import os
import warnings
from absl import flags
import matplotlib
from matplotlib import figure
from matplotlib.backends import backend_agg
import numpy as np
import tensorflow as tf
from tensorflow.contrib.learn.python.learn.datasets import mnist
import tensorflow_probability as tfp
from trainer.tfgfile_wrapper import tfgfile_wrapper
```

**Comment by gogasca** [comment]
@dizcology feedback provided. Do you still have plans to merge these samples?

**Comment by gogasca** [comment]
@dizcology friendly ping

**Comment by dizcology** [comment]
Closing this and moving to the ml-on-gcp repo.

**Pull request closed without merging**